pkg/util/kubernetes/apiserver/store.go

```diff
@@ -0,0 +1,79 @@
+package apiserver
+
+import (
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/patrickmn/go-cache"
+)
+
+type metaBundleStore struct {
+	mu      sync.RWMutex
+	cache   *cache.Cache
+	keyFunc func(...string) string
+}
+
+func (m *metaBundleStore) Get(nodeName string) (*MetadataMapperBundle, bool) {
+	cacheKey := m.keyFunc(metadataMapperCachePrefix, nodeName)
+
+	var metaBundle *MetadataMapperBundle
+
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	v, ok := m.cache.Get(cacheKey)
+	if !ok {
+		return nil, false
+	}
+
+	metaBundle, ok = v.(*MetadataMapperBundle)
+	if !ok {
+		log.Errorf("invalid cache format for the cacheKey: %s", cacheKey)
+		return nil, false
+	}
+
+	return metaBundle, true
+}
+
+func (m *metaBundleStore) GetOrCreate(nodeName string) *MetadataMapperBundle {
+	cacheKey := m.keyFunc(metadataMapperCachePrefix, nodeName)
+
+	var metaBundle *MetadataMapperBundle
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
```

Comment:
The lock is because we no have multiple goroutines that can set/get/delete **the same cache key**.

Reply:
Logic seems good to me.
The underlying issue that is solved is not the thread safe access to the cache but rather the concurrency.
If nodelistener submits an update at the same time as endpointlistener, it's important to not override the bundle